### ISSUE TITLE
with Jinja/DBT  Enforce {{ ref() }} or {{ source() }} - Do not allow un-decorated tables within .sql

### ISSUE THREAD

**Issue by corythorson-harbor** [sqlfluff/sqlfluff#6855]
### Search before asking

- [x] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### Description

Enhancement: Enforce {{ ref() }} or {{ source() }} in Jinja/DBT SQL Files

Add a SQLFluff rule to enforce the use of {{ ref() }} or {{ source() }} for table references in DBT .sql files, disallowing undecorated table names (e.g., schema.table). This ensures adherence to DBT best practices, improving model safety and maintainability. Includes support for Snowflake dialect and DBT templater.

Rule: Flag any table reference not wrapped in {{ ref() }} or {{ source() }}.
Config: Allow users to enable or disable via .sqlfluff.
Target: DBT users

### Use case

# Use Case: Enforcing `{{ ref() }}` or `{{ source() }}` in DBT SQL Files with SQLFluff

## Context
A data engineering team uses DBT to manage transformations in a Snowflake data warehouse, including snapshot tables. To ensure maintainability and prevent errors, the team follows DBT best practices, requiring all table references in `.sql` files to use `{{ ref() }}` or `{{ source() }}` instead of hard-coded table names (e.g., `schema.table`).

## Problem
Developers occasionally write DBT models with undecorated table names, leading to:
- **Fragility**: Hard-coded references break if schemas change.
- **Inconsistency**: Mixed reference styles reduce code readability.
- **Debugging Overhead**: Manual reviews to catch these errors slow down development.

## Solution
Implement a new SQLFluff rule (`dbt_ref_enforcement`) to lint DBT `.sql` files and enforce the use of `{{ ref() }}` or `{{ source() }}` for all table references, flagging undecorated table names as violations.

## Scenario
- **File**: `models/staging/snapshot_model.sql`
- **Incorrect Code**:
  ```sql
  SELECT * 
  FROM my_schema.my_table 
  ```
- **SQLFluff Output**:
  ```
  L:2 | P:10 | dbt_ref_enforcement | Table reference 'my_schema.my_table' must use {{ ref() }} or {{ source() }}.
  ```
- **Corrected Code**:
  ```sql
  SELECT * 
  FROM {{ ref('my_table') }} 
  ```

## Benefits
- **Consistency**: Ensures all DBT models use safe, portable table references.
- **Automation**: Identifies errors during CI/CD pipelines, reducing the need for manual reviews.
- **Scalability**: Supports large DBT projects with complex dependencies.
- **Snowflake Support**: Works seamlessly

### Dialect

# Dialect for SQLFluff DBT Ref Enforcement Enhancement

## Dialect
**Snowflake**

The enhancement targets DBT SQL files using the Snowflake dialect, ensuring that the `dbt_ref_enforcement` rule correctly parses and enforces `{{ ref() }}` or `{{ source() }}` for table references in Snowflake-compatible queries, such as those involving snapshot tables.

### Are you willing to work on and submit a PR to address the issue?

- [x] Yes I am willing to submit a PR!

### Code of Conduct

- [x] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)
